### ISSUE TITLE
Update WebSockets library

### ISSUE THREAD

**Issue by fassko** [apollographql/apollo-ios#1030]
Need to check before what's up with latest Starscream version because it has several issues.

**Comment by designatednerd** [comment]
@fassko You thinking Starscream 4.x here?

**Comment by fassko** [comment]
@designatednerd Yep Starscream 4.x

**Comment by fassko** [comment]
Seems that 4.x version is more stable and I will move with this in coming week(s) :)

**Comment by sneakyness** [comment]
Is there anything Starscream 4.x does that [URLSessionWebSocketTask](https://developer.apple.com/documentation/foundation/urlsessionwebsockettask) doesn't? Pending the update to that I think I might try to get a fork going that conditionally uses it if iOS 13.

**Comment by designatednerd** [comment]
My recollection is that Starscream is already doing that under the hood - @fassko you'd know better than me, can you confirm?

**Comment by fassko** [comment]
Yes, just need to release to cocoapods. SwiftPM should be fine. Just need to test it, just sadly I'm not using it on day-to-day development anymore myself.

**Comment by designatednerd** [comment]
Fiiiiinally upgraded with #1394

**Comment by designatednerd** [comment]
JK, undoing with #1395 - I need to have more time than I've got right now to get into what's happening with these changes under the hood, because they remove a bunch of stuff that we've been using and after a fairly significant amount of fighting I was able to get it to build, but not get it anywhere near passing tests.

**Comment by MycujooMats** [comment]
Any updates on this? We would love to use this library but we are currently stuck with a dependency conflict here.

**Comment by yayang** [comment]
Hi, I have met a socket issue, either.
sometimes the connect doesn't receive anything from server side,  but webpage and android works well.
In my app, 2 parts use socket to fetch server data, one works well, but the other works bad sometimes.
Is that the same problem with this issue?

And I have found that you said "Fiiiiinally upgraded with #1394", and the you have mentioned starscrem 4.x.
But I have checked in my app, the framework apollo-ios 0.36.0 is using starscream 3.1.1, I'm not sure why the starscrem 4.x is not the dependency of apollo-ios 0.36.0 as it has been released about 2 months.

**Comment by designatednerd** [comment]
Yep, the rollback was what "JK, undoing in 1395" in the next comment after woo hoo meant. 🤦‍♀️

Can you please open a separate ticket with more details about your issue, including what happens when it "works bad"? That's almost certainly a different problem.

**Comment by yayang** [comment]
> Yep, the rollback was what "JK, undoing in 1395" in the next comment after woo hoo meant. 🤦‍♀️
> 
> Can you please open a separate ticket with more details about your issue, including what happens when it "works bad"? That's almost certainly a different problem.

Thank you for your quick reply, I can't reproduce it now.
But still I think it may occur under some vpn.

One thing I do can reproduce is that, at the beginning of app launching time(about 1-2 seconds), the request sent by apollo-ios could not receive any response. So for now, I have to delay all request from apollo-ios to avoid the problem.

**Comment by designatednerd** [comment]
That, I've got no idea on. If you're able to create a more detailed example can you please open a separate issue? Thank you.

**Comment by jkastelicnov** [comment]
Hey all, thanks for keeping this project running. 

Just wanted to add that we are also locked into dependency conflict between apollo-ios and another library, over the starscream. 

I know that adding just another +1 isn't helpful; I'd consider contributing a PR, but given I don't really have much context on how Apollo uses websockets and that @designatednerd couldn't get this done despite a relatively large effort, I don't really feel qualified. 

Maybe we could start a branch with what was already done and I can try helping to push this thing forward.

**Comment by MatejBalantic** [comment]
Hey all, thanks for keeping this project running.

Just wanted to add that we are also locked into dependency conflict between apollo-ios and another library, over the starscream.

I know that adding just another +1 isn't helpful; I'd consider contributing a PR, but given I don't really have much context on how Apollo uses websockets and that @designatednerd couldn't get this done despite a relatively large effort, I don't really feel qualified.

Maybe we could start a branch with what was already done and I can try helping to push this thing forward.

**Comment by MatejBalantic** [comment]
For anyone in our situation: as a workaround, if you are not using the Apollo framework that relies on `starscream` (e.g. `ApolloWebSocket`) you can get away by forking the `apollo-ios`, bumping the `starscream` to `4.0` and then importing this fork as your dependency.

Obviously everything that is using `starscream` would be broken until this issue is closed, but at least you can unblock yourself if you are not actually using the Apollo frameworks that rely on this dependecy.

**Comment by designatednerd** [comment]
I would welcome a PR - my failure should not be interpreted as anything other than "I haven't worked with websockets much and I thought this was going to be a one-afternoon thing and got very frustrated when it became clear it wasn't because I have 37 million other things to do".

**Comment by designatednerd** [comment]
Took a second stab with #1659, if anyone on this thread has thoughts on the issue I ran into but couldn't figure out, would love to hear them!

**Comment by designatednerd** [comment]
Shipped (at last) with `0.42.0`

**Comment by MatejBalantic** [comment]
Thanks so much for your hard work and for figuring this one out, it really unblocked us. It's a bummer I couldn't be more helpful here, but I do hope we can contribute something back in the future.
